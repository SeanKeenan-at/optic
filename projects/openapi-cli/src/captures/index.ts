import { CapturedBody } from './body';
import { CapturedBodies } from './streams/captured-bodies';
import { OpenAPIV3 } from '../specs';

export { CapturedBody, CapturedBodies };
export { CapturedInteractions } from './streams/captured-interactions';

export interface CapturedInteraction {
  request: {
    host: string;
    method: OpenAPIV3.HttpMethods;
    path: string;
    body: CapturedBody;
    // TODO: add support for headers and query params
  };
  response: {
    statusCode: number;
    body: CapturedBody;
    // TODO: add support headers
  };
}
