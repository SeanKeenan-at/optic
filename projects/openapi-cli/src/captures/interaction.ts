import { CapturedBody } from './body';
import { OpenAPIV3 } from '../specs';
import { HttpArchive } from './streams/sources/har';
import { URL } from 'url';
import { HttpMethods } from '../operations';
import invariant from 'ts-invariant';

export interface CapturedInteraction {
  request: {
    host: string;
    method: OpenAPIV3.HttpMethods;
    path: string;
    body: CapturedBody | null;
    // TODO: add support for headers and query params
  };
  response: {
    statusCode: string;
    body: CapturedBody | null;
    // TODO: add support headers
  };
}
export class CapturedInteraction {
  static fromHarEntry(entry: HttpArchive.Entry): CapturedInteraction {
    const url = new URL(entry.request.url);

    const method = HttpMethods[entry.request.method];

    return {
      request: {
        host: url.hostname,
        method,
        path: url.pathname,
        body: null,
      },
      response: {
        statusCode: '' + entry.response.status,
        body: null,
      },
    };
  }
}

export type CapturedRequest = CapturedInteraction['request'];
export type CapturedResponse = CapturedInteraction['response'];
