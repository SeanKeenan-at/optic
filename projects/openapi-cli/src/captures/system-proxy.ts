import url from 'url';
import { createCommandFeedback } from '../commands/reporters/feedback';
import { platform, runCommand } from '../shell-utils';
import { parseMacNetworkLine } from './mac-system-proxy';

export class SystemProxy {
  constructor(
    private proxyUrl: string,
    private feedback: ReturnType<typeof createCommandFeedback>
  ) {}
  private stopCommand: (() => Promise<void>) | undefined = undefined;

  async start(device: string | undefined) {
    const { port } = url.parse(this.proxyUrl);

    if (platform === 'mac') {
      const name = await chooseInterfaceMac();

      const previousState = await Promise.all([
        runCommand(`networksetup -getwebproxy "${name}"`).then(
          parseMacNetworkLine
        ),
        runCommand(`networksetup -getsecurewebproxy "${name}"`).then(
          parseMacNetworkLine
        ),
        runCommand(`networksetup -getproxybypassdomains "${name}"`).then(
          (result) => result.trim()
        ),
      ]);

      await Promise.all([
        runCommand(`networksetup -setwebproxy "${name}" 127.0.0.1 ${port}`),
        runCommand(
          `networksetup -setsecurewebproxy "${name}" 127.0.0.1 ${port}`
        ),
        runCommand(
          `networksetup -setproxybypassdomains "${name}" "<-loopback>"`
        ),
      ]);

      this.feedback.notable(
        `Proxy running on ${this.proxyUrl}. System proxy updated`
      );

      this.stopCommand = async () => {
        // no proxy before
        if (!previousState[0].enabled && !previousState[1].enabled) {
          await Promise.all([
            runCommand(`networksetup -setwebproxystate "${name}" off`),
            runCommand(`networksetup -setsecurewebproxystate "${name}" off`),
          ]);
        } else {
          const nonSecureProxy = previousState[0];
          const secureProxy = previousState[1];
          const bypass = previousState[2];
          await Promise.all([
            runCommand(
              `networksetup -setwebproxystate "${name}" ${
                nonSecureProxy.host
              } ${nonSecureProxy.port} ${
                nonSecureProxy.authenticated ? 'on' : 'off'
              }`
            ),
            runCommand(
              `networksetup -setsecurewebproxystate "${name}" ${
                secureProxy.host
              } ${secureProxy.port} ${secureProxy.authenticated ? 'on' : 'off'}`
            ),
            runCommand(
              `networksetup -setproxybypassdomains "${name}" "${bypass}"`
            ),
          ]);
        }

        this.feedback.notable(`Mac System Proxy settings cleared`);
      };
    } else {
      this.feedback.notable(
        `Proxy running on ${this.proxyUrl}. System proxy updated`
      );
      console.log(
        `automatic proxy configuration is not supported on ${platform}`
      );
    }
  }
  async stop() {
    if (this.stopCommand) await this.stopCommand();
  }
}

export async function chooseInterfaceMac() {
  const network = await runCommand('scutil --nwi');
  const device = new RegExp(/Network interfaces: (en[0-9])/);

  const result = network.match(device);

  let devicePreference = result ? result[1] : 'en0';

  const hardware = await runCommand('networksetup -listallhardwareports');
  const regex = new RegExp(/(Hardware Port: (.*)\nDevice: (.*)\n)/);

  let name: string | undefined = undefined;
  hardware.split(/\n\s*\n/).forEach((entry) => {
    if (name) return;
    const match = regex.exec(entry);

    if (match && match[3] === devicePreference) {
      name = match[2];
    }
  });
  return name || 'Wi-Fi';
}
