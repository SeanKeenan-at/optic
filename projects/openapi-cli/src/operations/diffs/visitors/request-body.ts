import { CapturedRequest } from '../../../captures';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { OpenAPIV3 } from '../../../specs';
import MIMEType from 'whatwg-mimetype';

export function* visitRequestBody(
  request: CapturedRequest,
  spec?: OpenAPIV3.RequestBodyObject
): IterableIterator<OperationDiffResult> {
  if (!spec) {
    if (request.body) {
      // no request body was documented, but present in interaction
      yield {
        kind: OperationDiffResultKind.UnmatchedRequestBody,
        contentType: request.body.contentType,
      };
    }
    return; // no spec, nothing left to diff
  }

  if (!request.body) {
    if (spec.required) {
      yield {
        kind: OperationDiffResultKind.MissingRequestBody,
      };
    }

    return; // no request body, nothing left to diff
  }

  const requestBodyType = request.body.contentType
    ? new MIMEType(request.body.contentType)
    : null;

  const bodySpec =
    (requestBodyType &&
      Object.entries(spec.content).find(([rawType]) => {
        let parsed = new MIMEType(rawType);
        return parsed.essence == requestBodyType?.essence;
      })?.[1]) ||
    null;

  if (!bodySpec) {
    yield {
      kind: OperationDiffResultKind.UnmatchedRequestBody,
      contentType: request.body.contentType,
    };
  }
}
