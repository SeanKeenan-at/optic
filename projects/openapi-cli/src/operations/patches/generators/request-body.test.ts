import { requestBodyPatches } from './request-body';
import {
  diffInteractionByOperation,
  OperationDiffResultKind,
} from '../../diffs';
import { OpenAPIV3 } from '../../../specs';
import { Operation, HttpMethods } from '../../';
import { CapturedInteraction, CapturedBody } from '../../../captures';
import { OperationPatch } from '..';

describe('requestBodyPatches', () => {
  it('generates a patch for an unspecified request body', () => {
    let operationWithoutRequestBody = operationFixture(null);
    let interactionWithRequestBody = interactionFixture(
      CapturedBody.fromJSON({}, 'application/json')
    );

    let [diff] = [
      ...diffInteractionByOperation(
        interactionWithRequestBody,
        operationWithoutRequestBody
      ),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedRequestBody);

    let patches = [...requestBodyPatches(diff, operationWithoutRequestBody)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(
        patch,
        operationWithoutRequestBody
      ).expect('operation patch should apply to operation');

      expect(patchedOperation).toMatchSnapshot();
    }
  });

  it('generates a patch for a missing request body', () => {
    const operationWithRequestBody = operationFixture({
      content: {
        'application/json': {},
      },
      required: true,
    });

    const requestWithoutBody = interactionFixture(null);

    let [diff] = [
      ...diffInteractionByOperation(
        requestWithoutBody,
        operationWithRequestBody
      ),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.MissingRequestBody);

    let patches = [...requestBodyPatches(diff, operationWithRequestBody)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(
        patch,
        operationWithRequestBody
      ).expect('operation patch should apply to operation');

      expect(patchedOperation).toMatchSnapshot();
    }
  });
});

function operationFixture(
  requestBody: OpenAPIV3.RequestBodyObject | null
): Operation {
  return {
    requestBody,
    method: HttpMethods.POST,
    pathPattern: '/some-path',
    responses: {},
  };
}

function interactionFixture(
  requestBody: CapturedBody | null
): CapturedInteraction {
  return {
    request: {
      host: 'optic.test',
      method: HttpMethods.POST,
      path: '/some-path',
      body: requestBody,
    },
    response: {
      statusCode: '200',
      body: null,
    },
  };
}
