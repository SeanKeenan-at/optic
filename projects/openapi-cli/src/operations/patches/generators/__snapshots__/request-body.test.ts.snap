// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestBodyPatches generates a patch for a missing request body 1`] = `
Array [
  Object {
    "description": "make request body optional",
    "groupedOperations": Array [
      Object {
        "diff": Object {
          "op": "replace",
          "path": "/requestBody/required",
          "value": false,
        },
        "intent": "make request body optional",
        "operations": Array [],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`requestBodyPatches generates a patch for a missing request body 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "requestBody": Object {
    "content": Object {
      "application/json": Object {},
    },
    "required": true,
  },
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;

exports[`requestBodyPatches generates a patch for a unmatched request body content type 1`] = `
Array [
  Object {
    "description": "add 'text/csv' body as a valid request body type",
    "groupedOperations": Array [
      Object {
        "diff": Object {
          "op": "add",
          "path": "/requestBody/content/text~1csv",
          "value": Object {},
        },
        "intent": "add text/csv as content type",
        "operations": Array [],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`requestBodyPatches generates a patch for a unmatched request body content type 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "requestBody": Object {
    "content": Object {
      "application/json": Object {},
    },
    "required": true,
  },
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;

exports[`requestBodyPatches generates a patch for an unspecified request body 1`] = `
Array [
  Object {
    "description": "add 'application/json' body as a valid request body type",
    "groupedOperations": Array [
      Object {
        "diff": Object {
          "op": "add",
          "path": "/requestBody",
          "value": Object {
            "content": Object {},
          },
        },
        "intent": "add request body to operation",
        "operations": Array [],
      },
      Object {
        "diff": Object {
          "op": "add",
          "path": "/requestBody/content/application~1json",
          "value": Object {},
        },
        "intent": "add application/json as content type",
        "operations": Array [],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`requestBodyPatches generates a patch for an unspecified request body 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "requestBody": null,
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;
